kame/kame/dtcp/dtcpc.8
.\" Copyright (C) 1999 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     $Id: dtcpc.8,v 1.3 2000/05/27 13:55:40 itojun Exp $
.\"
.Dd Apr 28, 1999
.Dt DTCPC 8
.Os KAME
.\"
.Sh NAME
.Nm dtcpc
.Nd Dynamic Tunnel Configuration Protocol client
.\"
.Sh SYNOPSIS
.Nm dtcpc
.Op Fl d
.Op Fl i Ar if
.Op Fl p Ar port
.Op Fl r Ar routecommand
.Op Fl t Ar tuntype
.Op Fl u Ar user
.Ar server
.\"
.Sh DESCRIPTION
.Nm
implements Dynamic Tunnel Configuration Protocol
.Pq DTCP for short
proposed by Trumpet.
.Nm
will first ask the DTCP password to the user.
Then, the program will try to connect DTCP server on
.Ar server ,
to IPv4 TCP port 20200
.Po
or port specified by
.Ar port
.Pc ,
login as DTCP client, and maintain IPv6 over IPv4 tunnel connection.
.Pp
The spec defines several kind of tunnel requests.
.Nm
will request tunnel typed
.Ar tuntype
to the DTCP server.
The default value for
.Ar tuntype
is
.Dq Li tunnelonly .
For tunnel types, please refer to
.Xr dtcps 8 .
.Pp
By default,
.Nm
will use
.Li gif0
as tunnelling device.
This can be altered by using
.Fl i .
.Pp
.Fl r
specifies how the default route is configured.
If
.Dq Li static
is specified as
.Ar routecommand ,
.Nm
uses the
.Xr route 8
command to install the default route towards the corresponding
tunnel interface.
This is the default behavior.
If 
.Dq Li solicit
is specified,
.Nm
will call the
.Xr rtsol 8
command to send router solicitation messages toward the other
end of the tunnel.
.Pp
.Fl u
is needed when DTCP user account is different from
the login name on the host.
.Pp
.Fl d
puts
.Nm
to debug mode.
.\"
.Sh RETURN VALUES
.Nm
will exit with positive value on errors.
.\"
.Sh SEE ALSO
.Xr gif 4 ,
.Xr dtcps 8 ,
.Xr gifconfig 8
.\"
.Sh HISTORY
The
.Nm
command first appeared in WIDE/KAME IPv6 protocol stack kit.
.\"
.Sh BUGS
KAME kernel has some trouble handling
.Li /126
addresses.
To avoid trouble, please do not use
.Dq Fl t Li host
for the moment.
This should be fixed soon.
.Pp
.Nm
will not work if you are behind a NAT box
.Po
This is a protocol feature, not a bug
.Pc .


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/dtcp/dtcpc.rb
#! @PREFIX@/ruby

#
# dtcpc, Turmpet Dynamic Tunel Configuration Protocol client
#

#
# Copyright (C) 1999 WIDE Project.
# All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 3. Neither the name of the project nor the names of its contributors
#    may be used to endorse or promote products derived from this software
#    without specific prior written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.
#
# $Id: dtcpc.rb,v 1.4 2000/11/22 11:11:08 itojun Exp $
#

require "socket"
require "thread"
require "md5"

# XXX should be derived from system headers
IPPROTO_IPV6 = 41
IPV6_FAITH = 29
TUNIF = "gif"
TIMEOUT = 60
TUNTIMEOUT = 300
DEBUG = false

def usage()
  STDERR.print "usage: #{File.basename($0)} [-d] [-i if] [-p port] [-t tuntype] [-u username] server\n"
end

def seed()
  m = MD5.new(Time.now.to_s)
  m.update($$.to_s)
  m.update(Socket.gethostname())
  return m.digest.unpack("H32")[0].tr('a-f', 'A-F')
end

def authenticate(user, seed, pass)
  m = MD5.new(user)
  m.update(seed)
  m.update(pass)
  return m.digest.unpack("H32")[0].tr('a-f', 'A-F')
end

#------------------------------------------------------------

port = 20200
username = `whoami`
username.chomp!()
ousername = username
password = ''
intface = 'gif0'
tuntype = 'tunnelonly'
route = 'static'
$debug = DEBUG

# # test pattern
# challenge = '0B1517C87D516A5FA65BED722D51A04F'
# response = authenticate('foo', challenge, 'bar')
# if response == 'DAC487C8DFBBF9EE5C7F8CDCC37B62A3'
#   STDERR.print "good!\n"
# else
#   STDERR.print "something bad in authenticate()\n"
# end
# exit 0

while ARGV[0] =~ /^-/ do
  case ARGV[0]
  when /-d/
    $debug = !$debug
  when /-i/
    ARGV.shift
    intface = ARGV[0]
  when /-p/
    ARGV.shift
    port = ARGV[0]
  when /-r/
    ARGV.shift
    route = ARGV[0]
  when /-t/
    ARGV.shift
    tuntype = ARGV[0]
  when /-u/
    ARGV.shift
    username = ARGV[0]
  else
    usage()
    exit 0
  end
  ARGV.shift
end

if ARGV.length != 1
  usage()
  exit 1
end

dst = ARGV[0]

tty = open('/dev/tty', 'r')
system("stty -echo")
STDERR.print "password for #{username}: "
password = tty.readline
tty.close()
password.chomp!()
system("stty sane")
STDERR.print "\n"

res = []
t = Socket.getaddrinfo(dst, port, Socket::PF_INET, Socket::SOCK_STREAM, nil)
if (t.size <= 0)
  STDERR.print "FATAL: getaddrinfo failed (dst=#{dst} port=#{port})\n"
  exit 1
end
res += t

sockpool = []
names = []
listenthreads = []

s = nil
server = []
res.each do |i|
  begin
    s = TCPsocket.open(i[3], i[1])
  rescue
    next
  end
  server = i
  break
end

if server == []
  STDERR.print "could not connect to #{dst} port #{port}\n"
  exit 1
end

me = s.addr()[3]

STDERR.print "logging in to #{server[3]} port #{server[1]}\n"
# get greeting
t = s.readline
STDERR.print '>>', t if $debug
challenge = t.split(/ /)[1]

#STDERR.print "authenticate(#{username} #{challenge} #{password}): "
response = authenticate(username, challenge, password)
#STDERR.print response, "\n"
s.print "tunnel #{username} #{response} #{tuntype}\r\n"
STDERR.print ">>tunnel #{username} #{response} #{tuntype}\n" if $debug

t = s.readline
STDERR.print '>>', t if $debug
if (t =~ /^\+OK/)
  t.gsub!(/[\r\n]/, '')
  a = t.split(/ /)
  if me != a[1]
    STDERR.print "failed, you are behind a NAT box (#{me} != #{a[1]})\n"
    s.print "quit\r\n"
    s.shutdown(1)
    exit 1
  end
  STDERR.print "gifconfig #{intface} #{a[1]} #{a[2]}\n" if $debug
  system("gifconfig #{intface} #{a[1]} #{a[2]}")
  # global address for the tunnel is given
  if a.length == 5
    STDERR.print "ifconfig #{intface} inet6 #{a[3]} #{a[4]} prefixlen 126 alias\n" if $debug
    system("ifconfig #{intface} inet6 #{a[3]} #{a[4]} prefixlen 126 alias")
  end
  STDERR.print "ifconfig #{intface} up\n" if $debug
  system("ifconfig #{intface} up")
  STDERR.print "tunnel to #{a[2]} established.\n"
  if route == "static"
    STDERR.print "route add -inet6 default ::1" if $debug
    system("route add -inet6 default ::1")
    STDERR.print "route change -inet6 default -ifp #{intface}\n" if $debug
    system("route change -inet6 default -ifp #{intface}")
  end
  if route == "solicit"
    STDERR.print "rtsol #{intface}\n" if $debug
    system("rtsol #{intface}")
  end
  STDERR.print "default route was configured.\n"
  begin
    while TRUE
      STDERR.print "sleep(60)\n" if $debug
      sleep 60
      s.print "ping\r\n"
      STDERR.print ">>ping\n" if $debug
      t = s.readline
      STDERR.print '>>', t if $debug
    end
  ensure
    s.print "quit\r\n"
    s.shutdown(1)
    if a.length == 5
      STDERR.print "ifconfig #{intface} inet6 #{a[3]} #{a[4]} -alias\n" if $debug
      system("ifconfig #{intface} inet6 #{a[3]} #{a[4]} -alias")
    end
    STDERR.print "ifconfig #{intface} down\n" if $debug
    system("ifconfig #{intface} down")
    if route == "static"
      STDERR.print "route delete -inet6 default\n" if $debug
      system("route delete -inet6 default")
    end
    exit 0
  end
  exit 0
else
  t.gsub!(/[\r\n]*$/, '')
  STDERR.print "failed, reason: #{t}"
  s.print "quit\r\n"
  s.shutdown(1)
  exit 1
end
